### PR TITLE
Implement facet_search

### LINKED ISSUE #462: [v1.3] Search in facet values `facetSearch`
⚠️ This issue is generated, it means the examples and the namings do not necessarily correspond to the language of this repository.
Also, if you are a maintainer, please add any clarification and instructions about this issue.

Sorry if this is already wholly/partially implemented. Feel free to let me know about the state of this issue in the repo.

Related to https://github.com/meilisearch/integration-guides/issues/280

----

This issue is divided into two sections, first, you need to make the implementation, and second, you must update the code-samples (no one likes outdated docs, right?).

### New implementation

Related to:
- issue: https://github.com/meilisearch/meilisearch/issues/3611
- discussion: https://github.com/meilisearch/product/discussions/515#discussioncomment-6240970
- spec: https://github.com/meilisearch/specifications/pull/246

Users of bigger datasets could have difficulty interacting with vast lists of facets data since, until today, it was not possible to search on them.

To introduce this feature into the SDKs, we must:

Create a new method called `facetSearch(facetSearchQuery)`/`facet_search(facetSearchQuery)`
Besides all the other parameters that a regular `SearchQuery` can take. The new `FacetSearchQuery` can take a `facetName`, `facetQuery`, `filter`, `q`, `matchingStrategy`.

Params definition:
```yaml
facetName:
  type: string
  required: true
  example: '"genres"'
facetQuery:
  type: string
  example: '"Horror"'
q:
  type: string
  default: '""'
  example: '"Back to the future"'
matchingStrategy:
  type: string
  default: 'last'
filter: // https://www.meilisearch.com/docs/learn/fine_tuning_results/filtering
  type: array
  example: '"groupId = 1 AND genre = adventure"'
```

### Code samples

Inside of this file: `.code-samples.meilisearch.yml`:

1. Create a new entry with this key `facet_search_1` containing a call to the new `facetSearch` method.
Refer to this curl example and translate it by using the ruby methods: https://github.com/meilisearch/documentation/blob/e4fdc3d485c22e31a045b8a71bacf18b6a7d50e0/.code-samples.meilisearch.yaml#L1069

2. Add new code sample `facet_search_3` with this content:
Refer to this curl example and translate it by using the ruby methods: https://github.com/meilisearch/documentation/blob/e4fdc3d485c22e31a045b8a71bacf18b6a7d50e0/.code-samples.meilisearch.yaml#L1087

TODO:
- [ ] Add the new method `facetSearch`
- [ ] Update the code-samples accordingly
- [ ] Add integration tests

### PR DESCRIPTION
**Pull request by ellnix**
# Pull Request

## Related issue
Fixes #462

## PR checklist
Please check if your PR fulfills the following requirements:
- [X] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [X] Have you read the contributing guidelines?
- [X] Have you made sure that the title is accurate and descriptive of the changes?

### CODE REVIEW

File: lib/meilisearch/index.rb [hunk 1]
```diff
@@ -225,6 +225,7 @@ def delete_all_documents!
     #   attributes_to_search_on - Customize attributes to search on at search time.
     def search(query, options = {})
       attributes = { q: query.to_s }.merge(options.compact)
+      attributes.delete :matching_strategy
```
Comment (brunoocasali) [inline lib/meilisearch/index.rb:228]: I'm not sure if I miss something but I was not expecting this here...

Comment (ellnix) [reply lib/meilisearch/index.rb:228]: Oh lmao my bad. This is why you don't do git add . xD
That was just me testing if the specs break  when I remove the feature. I'll remove it.

Comment (ellnix) [reply lib/meilisearch/index.rb]: It should be removed now

**Comment by ellnix** [comment]
@brunoocasali 

Again the problem is that `class Index` is too long, but I cannot fix that without touching code that does not belong in this PR.

**Comment by ellnix** [comment]
Rebased, with linter fixed again. :rocket: :smile:

**Review by brunoocasali — APPROVED** [review]
Thanks a lot @ellnix 🎉 

bors merge

**Pull request merged**